{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook for manuscript\n",
    "# Trizzino, Park et al (2016)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Following scripts or their modified versions are used for RNA-seq data processing pipelines\n",
    "## 2015 - 2016"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "RNA-seq data processing\n",
    "Original scripts by MT, modified and additionally commented by YP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%writefile 'RNASEQ.PROCESSING.sh'\n",
    "\n",
    "#!/bin/bash\n",
    "\n",
    "# Marco Trizzino\n",
    "\n",
    "##########RNA-seq data processing\n",
    "\n",
    "##FOR EACH INDIVIDUAL SAMPLE WE HAD SEQUENCES FROM TWO LANES:\n",
    "\n",
    "#CALLING THE VARIABLES\n",
    "echo $1 # Read RNA-seq sample lane 1\n",
    "echo $2 # Read RNA-seq sample lane 2\n",
    "echo $3 # reference genome\n",
    "echo $4 # Sample output name \n",
    "echo $5 # rsem reference \n",
    "\n",
    "#PERFORMING QC OF THE SEQUENCES WITH FASTQC\n",
    "fastqc $1.fastq -o /OUTPUT_PATH\n",
    "fastqc $2.fastq -o /OUTPUT_PATH\n",
    "\n",
    "#TRIMMING ADAPTERS FROM RNA-seq reads\n",
    "trim_galore --cutadapt /cutadapt -stringency 5 -length 50 -o /OUTPUT_PATH/ -q 20 $1.fastq $2.fastq\n",
    "\n",
    "#STAR 2-PASS ALIGNMENT\n",
    "\n",
    "STAR --limitGenomeGenerateRAM 240000000000 \\\\\n",
    "--genomeDir $3 \\\\\n",
    "--readFilesIn $1_trimmed.fq \\\\\n",
    "--outSAMtype SAM \\\\\n",
    "--quantMode TranscriptomeSAM \\\\\n",
    "--outFilterMultimapNmax 10 \\\\\n",
    "--outFilterMismatchNmax 10 \\\\\n",
    "--outFilterMismatchNoverLmax 0.3 \\\\\n",
    "--alignIntronMin 21 \\\\\n",
    "--alignIntronMax 0 \\\\\n",
    "--alignMatesGapMax 0 \\\\\n",
    "--alignSJoverhangMin 5 \\\\\n",
    "--runThreadN 12 \\\\\n",
    "--twopassMode Basic \\\\\n",
    "--twopass1readsN 60000000 \\\\\n",
    "--sjdbOverhang 100 \\\\\n",
    "--outFileNamePrefix $4_lane1.\n",
    "\n",
    "STAR --limitGenomeGenerateRAM 240000000000 \\\\\n",
    "--genomeDir $3 \\\\\n",
    "--readFilesIn $2_trimmed.fq \\\\\n",
    "--outSAMtype SAM \\\\\n",
    "--quantMode TranscriptomeSAM \\\\\n",
    "--outFilterMultimapNmax 10 \\\\\n",
    "--outFilterMismatchNmax 10 \\\\\n",
    "--outFilterMismatchNoverLmax 0.3 \\\\\n",
    "--alignIntronMin 21 \\\\\n",
    "--alignIntronMax 0 \\\\\n",
    "--alignMatesGapMax 0 \\\\\n",
    "--alignSJoverhangMin 5 \\\\\n",
    "--runThreadN 12 \\\\\n",
    "--twopassMode Basic \\\\\n",
    "--twopass1readsN 60000000 \\\\\n",
    "--sjdbOverhang 100 \\\\\n",
    "--outFileNamePrefix $4_lane2.\n",
    "\n",
    "mv $4_lane1.Aligned.out.sam $4_lane1_STAR.sam\n",
    "mv $4_lane2.Aligned.out.sam $4_lane2_STAR.sam\n",
    "\n",
    "#CONVERTING SAM IN BAM\n",
    "samtools view -@ 8 -Sb $4_lane1_STAR.sam > $4_lane1_STAR.bam\n",
    "samtools view -@ 8 -Sb $4_lane2_STAR.sam > $4_lane2_STAR.bam\n",
    "\n",
    "#THE FOLLOWING COMMANDS SORT THE READS BASED ON CHROMOSOMES COORDINATES\n",
    "samtools sort -@ 8 -T $4_lane1_STAR.bam -O bam -o $4_lane1_STAR.bam $4_lane1_STAR_srt.bam \n",
    "samtools sort -@ 8 -T $4_lane2_STAR.bam -O bam -o $4_lane2_STAR.bam $4_lane2_STAR_srt.bam \n",
    "\n",
    "# merge BAM\n",
    "samtools merge -f -@ 8 $4_mgd_STAR_srt.bam $4_lane1_STAR_srt.bam $4_lane2_STAR_srt.bam\n",
    "\n",
    "# MAPQ filter step\n",
    "samtools view -@ 8 -q 10 -b $4_mgd_STAR_srt.bam > $4_mgd_STAR_srt_q10.bam\n",
    "\n",
    "#QUANTIFY GENE EXPRESSION WITH FEATURE COUNTS\n",
    "featureCounts -T 16 -C -s 1 -t exon \\\\\n",
    "-g gene_id -a /SPECIES_ENSEMBL_ANNOTATIONS/ \\\\\n",
    "-o /OUTPUT_PATH/$4_feature_Counts.txt $4_mgd_STAR_srt_q10.bam \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Orthologous gene identification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%writefile 'GENE.EXPRESSION.ANALYSIS.R'\n",
    "\n",
    "# Marco Trizzino\n",
    "\n",
    "################### FIND GENES ANNOTATED AS ORTHOLOGOUS IN THE SIX STUDIED PRIMATES, USING BIOMART \n",
    "#In R\n",
    "\n",
    "source(\"http://bioconductor.org/biocLite.R\")\n",
    "biocLite(\"biomaRt\") #363kb\n",
    "library(biomaRt)\n",
    "\n",
    "ensembl <-  useMart(\"ensembl\", dataset = \"hsapiens_gene_ensembl\")\n",
    "listDatasets(ensembl)\n",
    "listAttributes(ensembl)\n",
    "myFilter <- \"ensembl_gene_id\"\n",
    "myAttributes <- c(\"ensembl_gene_id\",\n",
    "                  \"ogarnettii_homolog_ensembl_gene\", \n",
    "                  \"ptroglodytes_homolog_ensembl_gene\",\n",
    "                  \"cjacchus_homolog_ensembl_gene\",\n",
    "                  \"tbelangeri_homolog_ensembl_gene\",\n",
    "                  \"mmurinus_homolog_ensembl_gene\",\n",
    "                  \"mmulatta_homolog_ensembl_gene\")\n",
    "#The following step is slow:\n",
    "GENEannot<- getBM(attributes =  myAttributes, \n",
    "                  filters =  myFilter,\n",
    "                  values =  rownames(Nreads), \n",
    "                  mart = ensembl)\n",
    "write.csv(file='ENSG_homologues.csv',GENEannot)\n",
    "sum(GENEannot==\"\")\n",
    "write.csv(file=\"ENSG_homologues_all.csv\",\n",
    "          GENEannot[GENEannot[,2]!=\"\" & GENEannot[,3]!=\"\"& GENEannot[,4]!=\"\"& GENEannot[,5]!=\"\"& GENEannot[,6]!=\"\"& GENEannot[,7]!=\"\",], \n",
    "          row.names=FALSE)\n",
    "\n",
    "\n",
    "###################### \n",
    "#PREPARE MATRIX OF RNA-SEQ counts INCLUDING THE 10,243 ORTHOLOGOUS GENEs FOR EACH SPECIMEN\n",
    "\n",
    "####FOR EACH INDIVIDUAL DO AS FOLLOWS (EXAMPLE WITH CHIMP CH_4X0430:\n",
    "Nreads <-read.delim(file=\"CH_4X0430_RNA_feature_Counts.txt\", header=T, row.names=NULL, sep=\"\\t\")\n",
    "names(Nreads)[7]=\"CH_4X0430\"\n",
    "NreadsCOUNTS <- Nreads[,c(1,7)]\n",
    "orthologs=read.csv(\"ENSG_homologues_all.csv\",  header = TRUE)\n",
    "index<-match(NreadsCOUNTS$Geneid,orthologs$ptroglodytes_homolog_ensembl_gene)\n",
    "NreadsCOUNTS$HumanID=orthologs$ensembl_gene_id[index]\n",
    "Counts_ortho= NreadsCOUNTS[!is.na(NreadsCOUNTS$HumanID),]\n",
    "write.csv(file=\"CH_4X0430_counts_ortho.csv\", Counts_ortho, row.names=FALSE)\n",
    "#After processing all of the individual, \n",
    "#merge the individual csv files in a single one \"GENE_COUNTS_ORTHOLOGOUS.csv\": \n",
    "#each file (= read counts for a give specimen) becomes a column\n",
    "\n",
    "###IMPORTANT: \n",
    "#Read counts need to be normalized by feature lengths before performing differential expression\n",
    "#Example for CH_4X0519\n",
    "lengths= read.csv(\"feature_legths_per_species_per_gene.csv\", \n",
    "                  header=TRUE, row.names=NULL) \n",
    "#open matrix with the feature lengths per each of the species for each of the orthologous genes\n",
    "lengths$CHIMP_factor <- lengths$HUMAN_lengths / lengths$CHIMP_lengths\n",
    "counts=read.csv(\"GENE_COUNTS_ORTHOLOGOUS.csv\", header = T)\n",
    "counts$CH_4X0519_norm <- counts$CH_4X0519*counts$CHfactor\n",
    "\n",
    "\n",
    "\n",
    "#### MATRIX WITH READ COUNTS FOR THE 10,683 ORTHOLOGOUS GENES, \n",
    "#NORMALIZED BY FEATURE LENGTHS FOR THE SIX SPECIES IS TABLE S3\n",
    "\n",
    "\n",
    "\n",
    "###################### \n",
    "#PERFORM DIFFERENTIAL EXPRESSION WITH DEseq2 \n",
    "#ON THE GENE COUNT MATRIX NORMALIZED BY FEATURE LENGTH (TABLE S3)\n",
    "\n",
    "#Download DESeq2 from Bioconductor (only need to do the first time you use DESeq2):\n",
    "source(\"http://bioconductor.org/biocLite.R\")\n",
    "biocLite(\"DESeq2\")\n",
    "\n",
    "#Load DESeq2- Do this step every time you open R and need to run DESeq2\n",
    "library(\"DESeq2\")\n",
    "\n",
    "#Import your expression matrix- mine was a comma seperated values file\n",
    "Nreads <-read.csv(\"GENE_COUNTS_ORTHOLOGOUS_NORM_FEAT_LENGTH.csv\", header=TRUE, row.names=1)\n",
    "#CRITICAL: You need to use read countsfor DESeq2, not normalized data like RPKMs (see manual).\n",
    "#If read counts are not integers, they need to be round up to integers:\n",
    "Nreads_norm=round(Nreads,0)\n",
    "#Omit any genes with zero counts across samples:\n",
    "Nreads <- Nreads[rowSums(Nreads)!=0,]\n",
    "#genes that have count=0 in only some of the genes need to be an integer >0\n",
    "Nreads=Nreads+1\n",
    "\n",
    "#Now import a data frame that provides annotatios for the samples \n",
    "#The rows of this correspond to your samples, \n",
    "#column 1 has the name of the sample and column two the condition (=species or species group)\n",
    "#for humans vs other primates:\n",
    "colData1 <-read.csv(\"colData_HSvsOTHER.csv\", header=TRUE, row.names=1) #HS VS ALL SPECIES\n",
    "#for apes vs other primates:\n",
    "colData2 <-read.csv(\"colData_APESvsALL.csv\", header=TRUE, row.names=1) #APES VS ALL SPECIES\n",
    "#for all species vs all species pair-wise comparisons:\n",
    "colData3 <-read.csv(\"colData_ALL_SPECIES.csv\", header=TRUE, row.names=1) #ALL SPECIES VS ALL SPECIES\n",
    "\n",
    "###Now I want to look at 'HS' vs 'OTHERS' no fold change, FDR=0.05\n",
    "#Perform DE analysis. You are telling it to compare by condition \n",
    "dds <- DESeqDataSetFromMatrix(countData = Nreads, colData=colData1)\n",
    "dds <- DESeq(dds)\n",
    "\n",
    "res_HS_OTHER<-results(dds, contrast=c(\"condition\",\"HS\",\"OTHER\"))\n",
    "#Remove those genes for which a multiple correction p-value is \"NA\"\n",
    "res_HS_OTHER_p=res_HS_OTHER[!is.na(res_HS_OTHER$padj), ] \n",
    "#Only keep genes with corrected p-values below 0.05:\n",
    "res_HS_OTHER_p=res_HS_OTHER_p[res_HS_OTHER_p$padj<0.05,] \n",
    "#Save logfile\n",
    "write.csv(as.data.frame(res_HS_OTHER_p), file = \"HSVsOTHER.csv\")\n",
    "#Get the names of upregulated genes (HS>OTHER):\n",
    "HighHS_OTHER=rownames(res_HS_OTHER_p[res_HS_OTHER_p$log2FoldChange>0,])\n",
    "write.csv(as.data.frame(HighHS_OTHER), file = \"HighHS_OTHER.csv\")\n",
    "#Get the names of downregulated genes (HS<OTHER):\n",
    "LowHS_OTHER=rownames(res_HS_OTHER_p[res_HS_OTHER_p$log2FoldChange<0,])\n",
    "write.csv(as.data.frame(LowHS_OTHER), file = \"LowHS_OTHER.csv\")\n",
    "\n",
    "\n",
    "###\n",
    "#Now I want to look at 'APES' vs 'OTHERS' no fold change, FDR=0.05\n",
    "dds <- DESeqDataSetFromMatrix(countData = Nreads, colData=colData2, design= ~condition)\n",
    "dds <- DESeq(dds)\n",
    "\n",
    "res_APES_OTHER<-results(dds, contrast=c(\"condition\",\"APES\",\"OTHER\"))\n",
    "#Remove those genes for which a multiple correction p-value is \"NA\"\n",
    "res_APES_OTHER_p=res_APES_OTHER[!is.na(res_APES_OTHER$padj), ] \n",
    "#write.csv(as.data.frame(res_APES_OTHER_p), file = \"APESVsOTHER_ALL_P.csv\")\n",
    "#Only keep genes with corrected p-values below 0.05:\n",
    "res_APES_OTHER_p=res_APES_OTHER_p[res_APES_OTHER_p$padj<0.05,] \n",
    "#Get the names of upregulated genes (APES>OTHER):\n",
    "HighAPES_OTHER=rownames(res_APES_OTHER_p[res_APES_OTHER_p$log2FoldChange>0,])\n",
    "write.csv(as.data.frame(HighAPES_OTHER), file = \"HighAPES_OTHER.csv\")\n",
    "#Get the names of downregulated genes (APES<OTHER):\n",
    "LowAPES_OTHER=rownames(res_APES_OTHER_p[res_APES_OTHER_p$log2FoldChange<0,])\n",
    "write.csv(as.data.frame(LowAPES_OTHER), file = \"LowAPES_OTHER.csv\")\n",
    "\n",
    "\n",
    "###Now all of the possible human vs any other species pair-wise combinations no fold change, FDR=0.05\n",
    "\n",
    "dds <- DESeqDataSetFromMatrix(countData = Nreads, colData=colData3, design= ~condition)\n",
    "dds <- DESeq(dds)\n",
    "\n",
    "# 'HS' vs 'BB'\n",
    "\n",
    "res_HS_BB<-results(dds, contrast=c(\"condition\",\"HS\",\"BB\"))\n",
    "#Remove those genes for which a multiple correction p-value is \"NA\"\n",
    "res_HS_BB_p=res_HS_BB[!is.na(res_HS_BB$padj), ] \n",
    "#Only keep genes with corrected p-values below 0.05:\n",
    "res_HS_BB_p=res_HS_BB_p[res_HS_BB_p$padj<0.05,] \n",
    "#Get the names of upregulated genes (HS>BB):\n",
    "HighHS_BB=rownames(res_HS_BB_p[res_HS_BB_p$log2FoldChange>0,])\n",
    "write.csv(as.data.frame(HighHS_BB), file = \"HighHS_BB.csv\")\n",
    "#Get the names of downregulated genes (HS<BB):\n",
    "LowHS_BB=rownames(res_HS_BB_p[res_HS_BB_p$log2FoldChange<0,])\n",
    "write.csv(as.data.frame(LowHS_BB), file = \"LowHS_BB.csv\")\n",
    "\n",
    "#'HS' vs 'CH' \n",
    "\n",
    "res_HS_CH<-results(dds, contrast=c(\"condition\",\"HS\",\"CH\"))\n",
    "#Remove those genes for which a multiple correction p-value is \"NA\"\n",
    "res_HS_CH_p=res_HS_CH[!is.na(res_HS_CH$padj), ] \n",
    "#Only keep genes with corrected p-values below 0.05:\n",
    "res_HS_CH_p=res_HS_CH_p[res_HS_CH_p$padj<0.05,] \n",
    "#Get the names of upregulated genes (HS>CH):\n",
    "HighHS_CH=rownames(res_HS_CH_p[res_HS_CH_p$log2FoldChange>0,])\n",
    "write.csv(as.data.frame(HighHS_CH), file = \"HighHS_CH.csv\")\n",
    "#Get the names of downregulated genes (HS<CH):\n",
    "LowHS_CH=rownames(res_HS_CH_p[res_HS_CH_p$log2FoldChange<0,])\n",
    "write.csv(as.data.frame(LowHS_CH), file = \"LowHS_CH.csv\")\n",
    "\n",
    "#'HS' vs 'RH' \n",
    "\n",
    "res_HS_RH<-results(dds, contrast=c(\"condition\",\"HS\",\"RH\"))\n",
    "#Remove those genes for which a multiple correction p-value is \"NA\"\n",
    "res_HS_RH_p=res_HS_RH[!is.na(res_HS_RH$padj), ] \n",
    "#Only keep genes with corrected p-values below 0.05:\n",
    "res_HS_RH_p=res_HS_RH_p[res_HS_RH_p$padj<0.05,] \n",
    "#Get the names of upregulated genes (HS>RH):\n",
    "HighHS_RH=rownames(res_HS_RH_p[res_HS_RH_p$log2FoldChange>0,])\n",
    "write.csv(as.data.frame(HighHS_RH), file = \"HighHS_RH.csv\")\n",
    "#Get the names of downregulated genes (HS<RH):\n",
    "LowHS_RH=rownames(res_HS_RH_p[res_HS_RH_p$log2FoldChange<0,])\n",
    "write.csv(as.data.frame(LowHS_RH), file = \"LowHS_RH.csv\")\n",
    "\n",
    "#'HS' vs 'MS'\n",
    "\n",
    "res_HS_MS<-results(dds, contrast=c(\"condition\",\"HS\",\"MS\"))\n",
    "#Remove those genes for whiMS a multiple correction p-value is \"NA\"\n",
    "res_HS_MS_p=res_HS_MS[!is.na(res_HS_MS$padj), ] \n",
    "#Only keep genes with corrected p-values below 0.05:\n",
    "res_HS_MS_p=res_HS_MS_p[res_HS_MS_p$padj<0.05,] \n",
    "#Get the names of upregulated genes (HS>MS):\n",
    "HighHS_MS=rownames(res_HS_MS_p[res_HS_MS_p$log2FoldChange>0,])\n",
    "write.csv(as.data.frame(HighHS_MS), file = \"HighHS_MS.csv\")\n",
    "#Get the names of downregulated genes (HS<MS):\n",
    "LowHS_MS=rownames(res_HS_MS_p[res_HS_MS_p$log2FoldChange<0,])\n",
    "write.csv(as.data.frame(LowHS_MS), file = \"LowHS_MS.csv\")\n",
    "\n",
    "#'HS' vs 'ML' \n",
    "\n",
    "res_HS_ML<-results(dds, contrast=c(\"condition\",\"HS\",\"ML\"))\n",
    "#Remove those genes for which a multiple correction p-value is \"NA\"\n",
    "res_HS_ML_p=res_HS_ML[!is.na(res_HS_ML$padj), ] \n",
    "#Only keep genes with corrected p-values below 0.05:\n",
    "res_HS_ML_p=res_HS_ML_p[res_HS_ML_p$padj<0.05,] \n",
    "#Get the names of upregulated genes (HS>ML):\n",
    "HighHS_ML=rownames(res_HS_ML_p[res_HS_ML_p$log2FoldChange>0,])\n",
    "write.csv(as.data.frame(HighHS_ML), file = \"HighHS_ML.csv\")\n",
    "#Get the names of downregulated genes (HS<ML):\n",
    "LowHS_ML=rownames(res_HS_ML_p[res_HS_ML_p$log2FoldChange<0,])\n",
    "write.csv(as.data.frame(LowHS_ML), file = \"LowHS_ML.csv\")\n",
    "\n",
    "#######################################################\n",
    "#DIFFERENTIAL EXPRESSION ANALYSIS RESULTS ARE IN TABLES4\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
